core/src/ledger_cleanup_service.rs

```diff
@@ -0,0 +1,101 @@
+//! The `ledger_cleanup_service` drops older ledger data to limit disk space usage
+
+use crate::blocktree::Blocktree;
+use crate::result::{Error, Result};
+use crate::service::Service;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::timing::DEFAULT_SLOTS_PER_EPOCH;
+use std::string::ToString;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{Receiver, RecvTimeoutError};
+use std::sync::Arc;
+use std::thread;
+use std::thread::{Builder, JoinHandle};
+use std::time::Duration;
+
+pub const DEFAULT_MAX_LEDGER_SLOTS: u64 = DEFAULT_SLOTS_PER_EPOCH * 5;
+
+pub struct LedgerCleanupService {
+    t_cleanup: JoinHandle<()>,
+}
+
+impl LedgerCleanupService {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(
+        slot_full_receiver: Receiver<(u64, Pubkey)>,
+        blocktree: Arc<Blocktree>,
+        max_ledger_slots: u64,
+        exit: &Arc<AtomicBool>,
+    ) -> Self {
+        let exit = exit.clone();
+        let t_cleanup = Builder::new()
+            .name("solana-ledger-cleanup".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+                if let Err(e) =
+                    Self::cleanup_ledger(&slot_full_receiver, &blocktree, max_ledger_slots)
+                {
+                    match e {
+                        Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
+                        Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
+                        _ => info!("Error from cleanup_ledger: {:?}", e),
+                    }
+                }
+            })
+            .unwrap();
+        Self { t_cleanup }
+    }
+
+    fn cleanup_ledger(
+        slot_full_receiver: &Receiver<(u64, Pubkey)>,
+        blocktree: &Arc<Blocktree>,
+        max_ledger_slots: u64,
+    ) -> Result<()> {
+        let (slot, _) = slot_full_receiver.recv_timeout(Duration::from_secs(1))?;
+        if slot > max_ledger_slots {
+            //cleanup
+            blocktree.delete_all_columns(0, Some(slot - max_ledger_slots));
```

Comment:
delete_all_columns() is a wrong name, conveys info about current implementation that might not survive a rewrite.

prune() isn't a wrong name, but is confusing if it takes 2 parameters.

how about transplant(), graft(), forget(), stipulate(), forswear(), forgo()? each of which would take a single parameter.

p.s. this looks scary

Reply:
Updated it to `purge_slots`. Not terribly attached to it so if you think something else works better I'll switch to that.